### PR TITLE
API-85: create a media

### PR DESCRIPTION
**Pull request by skeleton**
[//]: <> (<3 Thanks for taking the time to contribute! You're awesome! <3)

Create a media and associate it to a product. 

Content-Type is `multipart/form-data` to be able to upload a binary file.
Request has to be:
```
'file': binary file,
'product': {
   "identifier": "foo",
   "attribute": "a_file",
   "scope": null,
   "locale": null
}
```

[//]: <> (If you've never contributed to this repository before, please read https://github.com/akeneo/pim-community-dev/blob/master/.github/CONTRIBUTING.md)

**Description (for Contributor and Core Developer)**

[//]: <> (What does this Pull Request do? reference the related issue?)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added Behats                      | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php [hunk 1]
```diff
@@ -175,4 +238,131 @@ public function downloadAction(Request $request, $code)
...
+        try {
+            $this->productUpdater->update($product, $productValues);
+        } catch (PropertyException $e) {
+            $this->remover->remove($fileInfo);
+
+            throw new UnprocessableEntityHttpException($e->getMessage(), $e);
+        }
+
+        $violations = $this->validator->validate($product);
+        if ($violations->count() > 0) {
```
Comment (pierallard) [inline src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php]: Some other reviewers will say "yoda style" but personnally I don't care and it's more visible in this orientation.

File: src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php [hunk 2]
```diff
@@ -175,4 +238,131 @@ public function downloadAction(Request $request, $code)
...
+            throw new UnprocessableEntityHttpException('Property "file" is required.');
+        }
+
+        try {
+            $fileInfo = $this->fileStorer->store($files->get('file'), FileStorage::CATALOG_STORAGE_ALIAS, true);
+        } catch (FileTransferException $e) {
+            throw new UnprocessableEntityHttpException($e->getMessage(), $e);
+        } catch (FileRemovalException $e) {
+            throw new UnprocessableEntityHttpException($e->getMessage(), $e);
+        }
```
Comment (pierallard) [inline src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php]: Hopefully php7 will allow us to catch multiple exceptions classes :heart:

File: src/Pim/Bundle/ApiBundle/tests/integration/Controller/Media/CreateMediaFileIntegration.php [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    {
+        return $this->fileSystem->has($pathFile);
+    }
+
+    /**
+     * Remove all files generated by tests
+     *
+     * @param string $pathFile
+     */
+    protected function unlinkFiles($pathFile)
```
Comment (pierallard) [inline src/Pim/Bundle/ApiBundle/tests/integration/Controller/Media/CreateMediaFileIntegration.php]: unlinkFile

File: src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php (hunk 2 above)
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php]: php 7.1 :(

File: src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php [hunk 4]
```diff
@@ -175,4 +238,131 @@ public function downloadAction(Request $request, $code)
...
+    /**
+     * @param FileInfoInterface $fileInfo
+     * @param ProductInterface  $product
+     * @param array             $productInfos
+     *
+     * @throws HttpException
+     *
+     * @return ProductInterface
+     */
+    protected function processProduct(FileInfoInterface $fileInfo, ProductInterface $product, array $productInfos)
```
Comment (phaseinducer) [inline src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php]: linkFileToProduct ?

File: src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php [hunk 5]
```diff
@@ -175,4 +238,131 @@ public function downloadAction(Request $request, $code)
...
+    }
+
+    /**
+     * @param FileBag $files
+     *
+     * @throws HttpException
+     *
+     * @return FileInfoInterface
+     */
+    protected function processFile(FileBag $files)
```
Comment (phaseinducer) [inline src/Pim/Bundle/ApiBundle/Controller/MediaFileController.php]: storeFile ?

File: src/Pim/Bundle/ApiBundle/tests/integration/Controller/Media/CreateMediaFileIntegration.php [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+    }
+
+    /**
+     * Remove all files generated by tests
+     *
+     * @param string $pathFile
+     *
+     * @return bool
+     */
+    protected function isFileExists($pathFile)
```
Comment (phaseinducer) [inline src/Pim/Bundle/ApiBundle/tests/integration/Controller/Media/CreateMediaFileIntegration.php]: fileExists or doesFileExist

**Pull request merged**